File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/ScreenStructurePlugin.java
package com.scurab.android.anuitor.nanoplugin;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.View;

import com.scurab.android.anuitor.extract2.BaseExtractor;
import com.scurab.android.anuitor.extract2.BaseViewExtractorKt;
import com.scurab.android.anuitor.extract2.DetailExtractor;
import com.scurab.android.anuitor.reflect.ActivityThreadReflector;
import com.scurab.android.anuitor.reflect.WindowManager;
import com.scurab.android.anuitor.tools.HttpTools;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fi.iki.elonen.NanoHTTPD;

import static com.scurab.android.anuitor.tools.HttpTools.MimeType.APP_JSON;

/**
 * Created by jbruchanov on 24/06/2014.
 */
public class ScreenStructurePlugin extends BasePlugin {

    public static final String FILE = "screenstructure.json";
    public static final String PATH = "/" + FILE;

    private WindowManager mWindowManager;
    private ActivityThreadReflector mActivityThread;

    public ScreenStructurePlugin(WindowManager windowManager) {
        mWindowManager = windowManager;
        mActivityThread = new ActivityThreadReflector();
    }

    @Override
    public String[] files() {
        return new String[]{FILE};
    }

    @Override
    public String mimeType() {
        return HttpTools.MimeType.APP_JSON;
    }

    @Override
    public boolean canServeUri(String uri, File rootDir) {
        return PATH.equals(uri);
    }

    @SuppressWarnings("unchecked")
    @Override
    public NanoHTTPD.Response serveFile(String uri, Map<String, String> headers, NanoHTTPD.IHTTPSession session, File file, String mimeType) {
        String[] viewRootNames = mWindowManager.getViewRootNames();
        List<HashMap<String, Object>> resultDataSet = new ArrayList<>();

        for (String rootName : viewRootNames) {
            View v = mWindowManager.getRootView(rootName);
            Context c = v.getContext();
            HashMap<String, Object> data = new HashMap<>();
            data.put("RootName", rootName);
            resultDataSet.add(data);
            Activity activity = getActivity(v);
            if (activity == null && c instanceof Activity) {
                activity = (Activity) c;
            }

            if (activity != null) {
                fillActivity(activity, data);
            } else {
                BaseExtractor extractor = DetailExtractor.getExtractor(v);
                extractor.fillValues(v, data, null);
                if (c instanceof ContextWrapper) {
                    c = ((ContextWrapper) c).getBaseContext();
                    if (c instanceof Activity) {
                        HashMap<String, Object> sub = new HashMap<>();
                        data.put("OwnerActivity", sub);
                        fillActivity((Activity) c, sub);
                    }
                }
            }
            data.remove(BaseViewExtractorKt.OWNER);
        }

        Collections.reverse(resultDataSet);//stack order
        String json =  JSON.toJson(resultDataSet);
        NanoHTTPD.Response response = new OKResponse(APP_JSON, new ByteArrayInputStream(json.getBytes()));
        return response;
    }

    private void fillActivity(@Nullable Activity activity, HashMap<String, Object> data){
        if (activity != null) {
            BaseExtractor extractor = DetailExtractor.getExtractor(activity.getClass());
            extractor.fillValues(activity, data, null);
        }
    }

    @Nullable
    private Activity getActivity(@NonNull View view) {
        for (Activity activity : mActivityThread.getActivities()) {
            if (activity.getWindow().getDecorView().getRootView() == view.getRootView()) {
                return activity;
            }
        }
        return null;
    }
}

File: Android/service/src/main/java/com/scurab/android/anuitor/nanoplugin/ScreenStructurePlugin.kt
package com.scurab.android.anuitor.nanoplugin

import android.app.Activity
import android.content.ContextWrapper
import com.scurab.android.anuitor.Constants
import com.scurab.android.anuitor.extract2.DetailExtractor
import com.scurab.android.anuitor.extract2.ExtractingContext
import com.scurab.android.anuitor.reflect.ActivityThreadReflector
import com.scurab.android.anuitor.reflect.WindowManager
import com.scurab.android.anuitor.tools.HttpTools.MimeType.APP_JSON
import fi.iki.elonen.NanoHTTPD
import java.io.ByteArrayInputStream
import java.io.File
import java.util.*

private const val FILE_DEEP = "screenstructure.json"

class ScreenStructurePlugin(private val windowManager: WindowManager) : BasePlugin() {

    private val activityThread = ActivityThreadReflector()
    private val files = arrayOf(FILE_DEEP)
    private val path = "/$FILE_DEEP"

    override fun files(): Array<String> = files
    override fun mimeType(): String = APP_JSON
    override fun canServeUri(uri: String, rootDir: File): Boolean = path == uri

    override fun serveFile(uri: String, headers: Map<String, String>,
                           session: NanoHTTPD.IHTTPSession,
                           file: File,
                           mimeType: String): NanoHTTPD.Response {

        return try {
            val json = BasePlugin.JSON.toJson(deepStructure())
            OKResponse(APP_JSON, ByteArrayInputStream(json.toByteArray()))
        } catch (e: Throwable) {
            Response(NanoHTTPD.Response.Status.INTERNAL_ERROR,
                    NanoHTTPD.MIME_PLAINTEXT,
                    ByteArrayInputStream((e.message ?: "Null exception message").toByteArray()))
        }
    }

    private fun deepStructure(): List<Map<String, Any>> {
        val viewRootNames = windowManager.viewRootNames
        val resultDataSet = mutableListOf<Map<String, Any>>()

        for (rootName in viewRootNames) {
            val v = windowManager.getRootView(rootName)
            var c = v.context
            val data = TreeMap<String, Any>()
            data["RootName"] = rootName
            resultDataSet.add(data)
            var activity = activityThread.activities
                    .firstOrNull { it.window.decorView.rootView == v.rootView }
            if (activity == null && c is Activity) {
                activity = c
            }

            if (activity != null) {
                DetailExtractor.getExtractor(activity.javaClass).apply {
                    fillValues(activity, ExtractingContext(data))
                }
            } else {
                DetailExtractor
                        .getExtractor(v)
                        .fillValues(v, ExtractingContext(data))

                if (c is ContextWrapper) {
                    c = c.baseContext
                    if (c is Activity) {
                        val sub = TreeMap<String, Any>()
                        data["OwnerActivity"] = sub
                        DetailExtractor.getExtractor(c.javaClass).apply {
                            fillValues(c, ExtractingContext(sub))
                        }
                    }
                }
            }
            data.remove(Constants.OWNER)
        }

        resultDataSet.reverse()//stack order
        return resultDataSet
    }
}